[quickfix] Integration test is flaky

There appears to be a race condition in the startup of the integration test that causes the `beforeAll()` method of the integration test to fail to get a local repo:

https://github.com/bndtools/bnd/blob/b4cddedfe8f233e1005828d00f3c83a1345611bc/bndtools.core.test/src/bndtools/core/test/editors/quickfix/BuildpathQuickFixProcessorTest.java#L77-L91

This needs to be fixed.

> java.io.EOFException: No input provided for project

> It would be nice if we had a stack trace of that exception.

We often run into these exceptions when running integrations tests with parallel Maven build on Jenkins, see this [job](https://ci.openhab.org/job/openHAB-Core/).

I just ran a build with debug logging `-X` and this is the stacktrace showing:

```
[ERROR] Failed to execute goal biz.aQute.bnd:bnd-testing-maven-plugin:5.2.0:testing (default) on project org.openhab.core.model.item.tests: No input provided for project -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal biz.aQute.bnd:bnd-testing-maven-plugin:5.2.0:testing (default) on project org.openhab.core.model.item.tests: No input provided for project
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
Caused by: org.apache.maven.plugin.MojoExecutionException: No input provided for project
    at aQute.bnd.maven.testing.plugin.TestingMojo.execute (TestingMojo.java:160)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
Caused by: java.io.EOFException: No input provided for project
    at org.apache.tools.ant.Project.defaultInput (Project.java:1293)
    at org.apache.tools.ant.Project.demuxInput (Project.java:1315)
    at org.apache.tools.ant.DemuxInputStream.read (DemuxInputStream.java:73)
    at aQute.lib.io.IO.copy (IO.java:183)
    at aQute.libg.command.Command.execute (Command.java:167)
    at aQute.bnd.build.ProjectLauncher.launch (ProjectLauncher.java:374)
    at aQute.launcher.plugin.ProjectLauncherImpl.launch (ProjectLauncherImpl.java:197)
    at aQute.tester.junit.platform.plugin.ProjectTesterImpl.test (ProjectTesterImpl.java:99)
    at aQute.bnd.build.Project.test (Project.java:2520)
    at aQute.bnd.maven.testing.plugin.TestingMojo.lambda$getOperation$1 (TestingMojo.java:207)
    at aQute.bnd.maven.lib.resolve.BndrunContainer.execute (BndrunContainer.java:179)
    at aQute.bnd.maven.testing.plugin.TestingMojo.execute (TestingMojo.java:157)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)

```

The `aQute.lib.io.IO.copy` also sometimes has issues with `java.io.BufferedInputStream` in multi threaded Maven builds:

```
[ERROR] Failed to execute goal biz.aQute.bnd:bnd-testing-maven-plugin:5.2.0:testing (default) on project org.openhab.core.model.item.tests: Stream closed -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal biz.aQute.bnd:bnd-testing-maven-plugin:5.2.0:testing (default) on project org.openhab.core.model.item.tests: Stream closed
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
Caused by: org.apache.maven.plugin.MojoExecutionException: Stream closed
    at aQute.bnd.maven.testing.plugin.TestingMojo.execute (TestingMojo.java:160)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
Caused by: java.io.IOException: Stream closed
    at java.io.BufferedInputStream.getBufIfOpen (BufferedInputStream.java:176)
    at java.io.BufferedInputStream.read (BufferedInputStream.java:342)
    at aQute.lib.io.IO.copy (IO.java:183)
    at aQute.libg.command.Command.execute (Command.java:167)
    at aQute.bnd.build.ProjectLauncher.launch (ProjectLauncher.java:374)
    at aQute.launcher.plugin.ProjectLauncherImpl.launch (ProjectLauncherImpl.java:197)
    at aQute.tester.junit.platform.plugin.ProjectTesterImpl.test (ProjectTesterImpl.java:99)
    at aQute.bnd.build.Project.test (Project.java:2520)
    at aQute.bnd.maven.testing.plugin.TestingMojo.lambda$getOperation$1 (TestingMojo.java:207)
    at aQute.bnd.maven.lib.resolve.BndrunContainer.execute (BndrunContainer.java:179)
    at aQute.bnd.maven.testing.plugin.TestingMojo.execute (TestingMojo.java:157)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
```

It might be an idea to catch these exceptions in `IO.copy` when they occur?

--- a/.github/workflows/cibuild.yml
+++ b/.github/workflows/cibuild.yml
@@ -24,7 +24,6 @@ env:
     -Dhttp.keepAlive=false
     -Dmaven.wagon.http.pool=false
     -Dmaven.wagon.http.retryHandler.count=3
-  BNDTOOLS_CORE_TEST_NOJUNITOSGI: true
 
 defaults:
   run:

--- a/bndtools.core.test/src/bndtools/core/test/editors/quickfix/AbstractBuildpathQuickFixProcessorTest.java
+++ b/bndtools.core.test/src/bndtools/core/test/editors/quickfix/AbstractBuildpathQuickFixProcessorTest.java
@@ -24,8 +24,6 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.TimeUnit;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -41,12 +39,7 @@ import org.assertj.core.presentation.StandardRepresentation;
 import org.eclipse.core.resources.IFolder;
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.resources.ResourcesPlugin;
-import org.eclipse.core.runtime.jobs.Job;
-import org.eclipse.jdt.core.ElementChangedEvent;
 import org.eclipse.jdt.core.ICompilationUnit;
-import org.eclipse.jdt.core.IElementChangedListener;
-import org.eclipse.jdt.core.IJavaElement;
-import org.eclipse.jdt.core.IJavaElementDelta;
 import org.eclipse.jdt.core.IJavaProject;
 import org.eclipse.jdt.core.IPackageFragment;
 import org.eclipse.jdt.core.IPackageFragmentRoot;
@@ -132,17 +125,24 @@ abstract class AbstractBuildpathQuickFixProcessorTest {
 
 		Path bundleRoot = Paths.get(System.getProperty("bndtools.core.test.dir"))
 			.resolve("./generated/");
+		log("Attempting to import fodder bundles from " + bundleRoot);
 		Files.walk(bundleRoot, 1)
 			.filter(x -> x.getFileName()
 				.toString()
 				.contains(".fodder."))
 			.forEach(bundle -> {
 				try {
+					log("Adding fodder bundle to localRepo: " + bundle);
 					localRepo.put(IO.stream(bundle), null);
 				} catch (Exception e) {
 					throw Exceptions.duck(e);
 				}
 			});
+		TaskUtils.log(() -> ("Local Index contains:\n\t" + localRepo.list("*")
+			.stream()
+			.collect(Collectors.joining(",\n\t"))));
+		TaskUtils.updateWorkspace("beforeAllBase()");
+
 		initSUTClass();
 
 		eclipseProject = ResourcesPlugin.getWorkspace()
@@ -170,17 +170,7 @@ abstract class AbstractBuildpathQuickFixProcessorTest {
 		synchronously("createPackageFragment", monitor -> pack = root.createPackageFragment("test", false, monitor));
 
 		// Wait for build to finish - attempted fix for #4553.
-		//
-		// Note that there is a possible race condition here as I'm not sure if
-		// it's guaranteed that the build would have started yet; if we start
-		// trying to wait for it before it actually starts it might return
-		// straight away instead of waiting for the finish. Need to think of a
-		// way to make sure we don't start waiting for it to finish until we
-		// know that it's already started.
-		Job.getJobManager()
-			.join(ResourcesPlugin.FAMILY_MANUAL_BUILD, null);
-		Job.getJobManager()
-			.join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+		TaskUtils.waitForBuild("beforeAllBase()");
 	}
 
 	protected static final String DEFAULT_CLASS_NAME = "Test";
@@ -190,35 +180,6 @@ abstract class AbstractBuildpathQuickFixProcessorTest {
 		locs = null;
 		problems = null;
 		problem = null;
-	}
-
-	static void waitForClasspathUpdate() throws Exception {
-		waitForClasspathUpdate(null);
-	}
-
-	static void waitForClasspathUpdate(String context) throws Exception {
-		String prefix = context == null ? "" : context + ": ";
-		CountDownLatch flag = new CountDownLatch(1);
-		IElementChangedListener listener = new ClasspathChangedListener(flag);
-		JavaCore.addElementChangedListener(listener, ElementChangedEvent.POST_CHANGE);
-		try {
-			// This call to forceRefresh() here is unnecessary if you are adding
-			// to
-			// the buildpath, but it seems to be necessary when clearing the
-			// build
-			// path. Without it, the Project object seems to hang around without
-			// updating and doesn't update its buildpath setting.
-			bndProject.forceRefresh();
-			Central.refreshFile(bndProject.getPropertiesFile());
-			log(prefix + "waiting for classpath to update");
-			if (flag.await(10000, TimeUnit.MILLISECONDS)) {
-				log(prefix + "done waiting for classpath to update");
-			} else {
-				log(prefix + "WARNING: timed out waiting for classpath to update");
-			}
-		} finally {
-			JavaCore.removeElementChangedListener(listener);
-		}
 	}
 
 	static void clearBuildpath() {
@@ -231,7 +192,8 @@ abstract class AbstractBuildpathQuickFixProcessorTest {
 				model.setBuildPath(Collections.emptyList());
 				model.saveChanges();
 				Central.refresh(bndProject);
-				waitForClasspathUpdate("clearBuildpath()");
+				TaskUtils.requestClasspathUpdate("clearBuildpath()");
+				TaskUtils.waitForBuild("clearBuildpath()");
 			} else {
 				log("buildpath was not set; not trying to clear it");
 			}
@@ -250,51 +212,10 @@ abstract class AbstractBuildpathQuickFixProcessorTest {
 			}
 			model.saveChanges();
 			Central.refresh(bndProject);
-			waitForClasspathUpdate("addBundleToBuildpath");
+			TaskUtils.requestClasspathUpdate("addBundleToBuildpath()");
+			TaskUtils.waitForBuild("addBundleToBuildpath()");
 		} catch (Exception e) {
 			throw Exceptions.duck(e);
-		}
-	}
-
-	// Listener used for synchronizing on classpath changes
-	static class ClasspathChangedListener implements IElementChangedListener {
-
-		final CountDownLatch flag;
-
-		ClasspathChangedListener(CountDownLatch flag) {
-			this.flag = flag;
-		}
-
-		@Override
-		public void elementChanged(ElementChangedEvent event) {
-			visit(event.getDelta());
-		}
-
-		private void visit(IJavaElementDelta delta) {
-			IJavaElement el = delta.getElement();
-			switch (el.getElementType()) {
-				case IJavaElement.JAVA_MODEL :
-					visitChildren(delta);
-					break;
-				case IJavaElement.JAVA_PROJECT :
-					if (isClasspathChanged(delta.getFlags())) {
-						flag.countDown();
-					}
-					break;
-				default :
-					break;
-			}
-		}
-
-		private boolean isClasspathChanged(int flags) {
-			return 0 != (flags
-				& (IJavaElementDelta.F_CLASSPATH_CHANGED | IJavaElementDelta.F_RESOLVED_CLASSPATH_CHANGED));
-		}
-
-		public void visitChildren(IJavaElementDelta delta) {
-			for (IJavaElementDelta c : delta.getAffectedChildren()) {
-				visit(c);
-			}
 		}
 	}
 

--- a/bndtools.core.test/src/bndtools/core/test/utils/TaskUtils.java
+++ b/bndtools.core.test/src/bndtools/core/test/utils/TaskUtils.java
@@ -4,14 +4,25 @@ import java.lang.reflect.InvocationTargetException;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 
+import org.bndtools.api.BndtoolsConstants;
 import org.eclipse.core.resources.IResourceVisitor;
 import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.resources.IWorkspaceDescription;
+import org.eclipse.core.resources.IncrementalProjectBuilder;
 import org.eclipse.core.resources.ResourcesPlugin;
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.NullProgressMonitor;
 import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.jdt.core.ClasspathContainerInitializer;
+import org.eclipse.jdt.core.IJavaModel;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
 
+import aQute.bnd.build.Workspace;
 import aQute.bnd.exceptions.Exceptions;
+import aQute.bnd.exceptions.SupplierWithException;
+import bndtools.central.Central;
 
 public class TaskUtils {
 
@@ -23,8 +34,16 @@ public class TaskUtils {
 
 	private TaskUtils() {}
 
+	public static void log(SupplierWithException<String> msg) {
+		try {
+			log(msg.get());
+		} catch (Exception e) {
+			throw Exceptions.duck(e);
+		}
+	}
+
 	public static void log(String msg) {
-		// System.err.println(System.currentTimeMillis() + ": " + msg);
+	//	System.err.println(System.currentTimeMillis() + ": " + msg);
 	}
 
 	public static void synchronously(String msg, MonitoredTask task) {
@@ -62,10 +81,75 @@ public class TaskUtils {
 	public static void dumpWorkspace() {
 		IWorkspace ws = ResourcesPlugin.getWorkspace();
 		try {
-		ws.getRoot()
-			.accept(VISITOR);
+			ws.getRoot()
+				.accept(VISITOR);
 		} catch (CoreException e) {
 			throw Exceptions.duck(e);
 		}
+	}
+
+	public static boolean setAutobuild(boolean on) {
+		try {
+			IWorkspace eclipse = ResourcesPlugin.getWorkspace();
+			IWorkspaceDescription description = eclipse.getDescription();
+			boolean original = description.isAutoBuilding();
+			description.setAutoBuilding(on);
+			eclipse.setDescription(description);
+			return original;
+		} catch (CoreException e) {
+			throw Exceptions.duck(e);
+		}
+	}
+
+	// Synchronously build the workspace
+	public static void buildFull() {
+		try {
+			ResourcesPlugin.getWorkspace()
+				.build(IncrementalProjectBuilder.FULL_BUILD, new NullProgressMonitor());
+		} catch (CoreException e) {
+			throw Exceptions.duck(e);
+		}
+	}
+
+	public static void buildIncremental() {
+		try {
+			ResourcesPlugin.getWorkspace()
+				.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, new NullProgressMonitor());
+		} catch (CoreException e) {
+			throw Exceptions.duck(e);
+		}
+	}
+
+	public static void updateWorkspace(String context) throws Exception {
+		log(context + ": Updating Bnd workspace");
+		Workspace bndWs = Central.getWorkspace();
+		bndWs.clear();
+		bndWs.forceRefresh();
+		bndWs.getPlugins();
+	}
+
+	public static void requestClasspathUpdate(String context) {
+		try {
+			log(context + ": Initiating classpath update");
+			ClasspathContainerInitializer initializer = JavaCore
+				.getClasspathContainerInitializer(BndtoolsConstants.BND_CLASSPATH_ID.segment(0));
+			if (initializer != null) {
+				IJavaModel javaModel = JavaCore.create(ResourcesPlugin.getWorkspace()
+					.getRoot());
+
+				for (IJavaProject project : javaModel.getJavaProjects()) {
+					initializer.requestClasspathContainerUpdate(BndtoolsConstants.BND_CLASSPATH_ID, project, null);
+				}
+			}
+		} catch (CoreException e) {
+			throw Exceptions.duck(e);
+		}
+		log(context + ": done classpath update");
+	}
+
+	public static void waitForBuild(String context) throws InterruptedException {
+		log(context + ": Initiating build");
+		buildIncremental();
+		log(context + ": done waiting for build to complete");
 	}
 }

--- a/bndtools.core.test/src/bndtools/core/test/utils/WorkbenchExtension.java
+++ b/bndtools.core.test/src/bndtools/core/test/utils/WorkbenchExtension.java
@@ -37,6 +37,7 @@ public class WorkbenchExtension implements BeforeAllCallback {
 
 	@Override
 	public void beforeAll(ExtensionContext context) throws Exception {
+		TaskUtils.setAutobuild(false);
 		BundleContext bc = FrameworkUtil.getBundle(context.getRequiredTestClass())
 			.getBundleContext();
 

--- a/bndtools.core.test/src/bndtools/core/test/utils/WorkspaceImporter.java
+++ b/bndtools.core.test/src/bndtools/core/test/utils/WorkspaceImporter.java
@@ -2,25 +2,18 @@ package bndtools.core.test.utils;
 
 import java.lang.reflect.InvocationTargetException;
 import java.nio.file.Path;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.TimeUnit;
 import java.util.stream.Stream;
 
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.resources.IWorkspace;
 import org.eclipse.core.resources.IWorkspaceRoot;
 import org.eclipse.core.resources.ResourcesPlugin;
-import org.eclipse.core.resources.WorkspaceJob;
 import org.eclipse.core.runtime.IProgressMonitor;
-import org.eclipse.core.runtime.IStatus;
-import org.eclipse.core.runtime.Status;
-import org.eclipse.core.runtime.jobs.Job;
 import org.eclipse.ui.dialogs.IOverwriteQuery;
 import org.eclipse.ui.wizards.datatransfer.FileSystemStructureProvider;
 import org.eclipse.ui.wizards.datatransfer.ImportOperation;
 
 import aQute.bnd.exceptions.Exceptions;
-import bndtools.central.Central;
 
 public class WorkspaceImporter {
 	private static final IOverwriteQuery	overwriteQuery	= file -> IOverwriteQuery.ALL;
@@ -37,14 +30,17 @@ public class WorkspaceImporter {
 
 	public void reimportProject(String projectName) {
 		try {
+			IWorkspace ws = ResourcesPlugin.getWorkspace();
 			IWorkspaceRoot wsr = ResourcesPlugin.getWorkspace()
 				.getRoot();
 
-			IProject project = wsr.getProject(projectName);
-			if (project.exists()) {
-				project.delete(true, true, null);
-			}
-			importProject(root.resolve(projectName), null);
+			ws.run(monitor -> {
+				IProject project = wsr.getProject(projectName);
+				if (project.exists()) {
+					project.delete(true, true, null);
+				}
+				importProject(root.resolve(projectName), null);
+			}, null);
 		} catch (Exception e) {
 			throw Exceptions.duck(e);
 		}
@@ -62,41 +58,21 @@ public class WorkspaceImporter {
 		IWorkspace ws = ResourcesPlugin.getWorkspace();
 		IWorkspaceRoot wsr = ResourcesPlugin.getWorkspace()
 			.getRoot();
-		Job job = new WorkspaceJob("Clean and import all") {
-			@Override
-			public IStatus runInWorkspace(IProgressMonitor monitor) {
-				try {
-					// Clean the workspace
-					IProject[] existingProjects = wsr.getProjects();
-					for (IProject project : existingProjects) {
-						project.delete(true, true, monitor);
-					}
-
-					projects.forEach(path -> importProject(path, monitor));
-					return Status.OK_STATUS;
-				} catch (Exception e) {
-					return new Status(IStatus.ERROR, WorkspaceImporter.class, 0,
-						"Error during import: " + e.getMessage(), e);
-				}
-			}
-		};
-		// Lock the entire workspace so that we don't run simultaneously with
-		// other jobs; see eg #4573.
-		job.setRule(wsr);
-		job.schedule();
 		try {
-			if (!job.join(10000, null)) {
-				TaskUtils.dumpWorkspace();
-				throw new IllegalStateException("Timed out waiting for workspace import to complete");
-			}
+			ws.run(monitor -> {
+				// Clean the workspace
+				IProject[] existingProjects = wsr.getProjects();
+				for (IProject project : existingProjects) {
+					project.delete(true, true, monitor);
+				}
 
-			// Wait for Workspace object to be complete.
-			final CountDownLatch flag = new CountDownLatch(1);
-			Central.onCnfWorkspace(bndWS -> {
-				flag.countDown();
-			});
-			flag.await(10000, TimeUnit.MILLISECONDS);
-		} catch (InterruptedException e) {
+				projects.forEach(path -> importProject(path, monitor));
+			}, null);
+
+			TaskUtils.updateWorkspace("importAllProjects()");
+			TaskUtils.requestClasspathUpdate("importAllProjects()");
+			TaskUtils.waitForBuild("importAllProjects()");
+		} catch (Exception e) {
 			throw Exceptions.duck(e);
 		}
 	}

